Integrate UnifiedHighlighter

jimczi opened the pull request:
This change integrates the Lucene highlighter called "unified" in the list of supported highlighters for ES.
This highlighter has multiple modes:
  * plain: a mode that analyzes the plain text directly
  * postings: a mode that uses the postings offsets to perform the highlight
  * fvh: a mode that uses the term vectors to perform the highlighting

Since this is a "unified" highlighter here is the complete list of highlighting features supported or not by this integration:

- [x] works on fields with `term_vectors`, `offsets` or simply`stored`(or extracted from _source).
- [x] `force_source` 
- [x] `pre_tags` and `post_tags`
Multiple tags are not allowed
- [x] `encoder` (html and default)
- [x] `number_of_fragments`
- [x] `no_match_size`
- [x] `require_field_match`
Requires http://issues.apache.org/jira/browse/LUCENE-7575. Note that LUCENE-7575 is more than just require_field_match since it can select which fields to highlight
- [ ] `matched_fields` 
- [ ] `fragment_size ` 
Requires http://issues.apache.org/jira/browse/LUCENE-7565
- [ ] `boundary_chars`
- [ ] `boundary_max_scan`
- [x] Queries with rewrite that needs an index reader are ignored (MultiPhrasePrefixQuery, CommonTermsQuery, AllFieldQuery, ...).
- [ ] Collapse contiguous highlights: `<b>a</b><b>b</b><b>c</b>` => `<b>abc</b>`

Fixes #21376

jimczi commented:
Documentation is missing (which is why this is a WIP ;) )

jpountz commented on core/src/main/java/org/apache/lucene/search/uhighlight/CustomPassageFormatter.java:
```diff
@@ -0,0 +1,82 @@
+        return snippets;
+    }
+
+    protected void append(StringBuilder dest, String content, int start, int end) {
```
Can you write some javadoc as to why a custom impl would want to override this?

jpountz commented on core/src/main/java/org/elasticsearch/index/search/MatchQuery.java:
```diff
@@ -328,6 +329,12 @@ public Query createPhrasePrefixQuery(String field, String queryText, int phraseS
+                if (bq.getQuery() instanceof TermQuery) {
+                    prefixQuery.add(((TermQuery) bq.getQuery()).getTerm());
+                    return prefixQuery;
+                }
```
this looks like a different change?

jpountz commented on core/src/main/java/org/elasticsearch/search/SearchModule.java:
```diff
@@ -615,7 +616,11 @@ public void register(String name, Suggester<?> t) {
+        highlighters.register("unified", uh);
+        highlighters.register("unified_plain", uh);
+        highlighters.register("unified_postings", uh);
+        highlighters.register("unified_fvh", uh);
```
why do we need to register the same instance under multiple names?

jpountz replied:
oh I see then we have a switch on the highlighter type in the impl, maybe write a comment about it here?

nik9000 replied:
The experimental highlighter used another field if you were going to override. It called it `"hit_source"` I think. I like that kind of thing better because it makes it more clear that you are only changing where the hits are calculated from not the actual highlighter implementation.

jpountz commented on core/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/UnifiedHighlighter.java:
```diff
@@ -0,0 +1,167 @@
+
+        if (field.fieldOptions().scoreOrdered()) {
+            //let's sort the snippets by score if needed
+            CollectionUtil.introSort(snippets, (o1, o2) -> (int) Math.signum(o2.getScore() - o1.getScore()));
```
maybe use Double.compare rather than Math.signum?

javanna commented on core/src/main/java/org/apache/lucene/search/postingshighlight/CustomPostingsHighlighter.java:
```diff
@@ -22,6 +22,7 @@
 import org.apache.lucene.analysis.Analyzer;
 import org.apache.lucene.search.IndexSearcher;
 import org.apache.lucene.search.Query;
+import org.apache.lucene.search.highlight.Snippet;
```
leftover?

javanna commented on core/src/main/java/org/apache/lucene/search/uhighlight/CustomPassageFormatter.java:
```diff
@@ -0,0 +1,82 @@
+1) extract different snippets (instead of a single big string) together with their scores ({@link Snippet})
+2) use the {@link Encoder} implementations that are already used with the other highlighters
+ */
+public class CustomPassageFormatter extends PassageFormatter {
```
is there no way to re-use the already existing CustomPassageFormatter class?

javanna commented on core/src/main/java/org/apache/lucene/search/uhighlight/CustomPassageFormatter.java:
```diff
@@ -0,0 +1,82 @@
+ * under the License.
+ */
+
+package org.apache.lucene.search.uhighlight;
```
I am not a big fan of `uhighlight` as a package name, maybe we can simply use "unifiedhighlight", since we already have "postingshighlight"

Timothy055 replied:
We had similar thoughts (but were instead considering unifiedhighlighter as a package).  Still it seemed too long.

dsmiley replied:
Feel free to file an issue to LUCENE to rename the package to "unifiedhighlight".  It's labelled "@lucene.experimental" and was just released.

javanna commented on core/src/test/java/org/apache/lucene/search/postingshighlight/CustomPostingsHighlighterTests.java:
```diff
@@ -31,6 +31,7 @@
 import org.apache.lucene.index.Term;
 import org.apache.lucene.search.IndexSearcher;
 import org.apache.lucene.search.Query;
+import org.apache.lucene.search.highlight.Snippet;
```
leftover?

javanna commented on core/src/test/java/org/apache/lucene/search/postingshighlight/CustomPassageFormatterTests.java:
```diff
@@ -19,6 +19,7 @@
 
 package org.apache.lucene.search.postingshighlight;
 
+import org.apache.lucene.search.highlight.Snippet;
```
leftover?

javanna commented:
I think docs are super important. We already have 3 highlighters and it is hard enough for users to figure out which one to use. Would be nice to come up with reasons to prefer this new implementation over the existing ones etc.

nik9000 commented:
> Would be nice to come up with reasons to prefer this new implementation over the existing ones etc.

I expect this one allows you to source the hits from postings while still using fvh-like snippet segmentation. So it works properly on more free form fields. I expect this but I have not confirmed it to be so.

jimczi commented:
> I expect this one allows you to source the hits from postings while still using fvh-like snippet segmentation. So it works properly on more free form fields. I expect this but I have not confirmed it to be so.

Well it's the other way around ;). The unified highlighter is a fork of the postings highlighter that can also works with term vectors or plain analysis. The snippet segmentation is exactly the same as the postings one. I only see this highlighter as a generalization of the postings highlighter. After some benchmarking it seems to perform quite well on plain analysis (twice faster than the original plain highlighter on wikipedia text) but the drawbacks are the same than for the original Postings Highlighter.

nik9000 commented:
> Well it's the other way around ;). The unified highlighter is a fork of the postings highlighter that can also works with term vectors or plain analysis. The snippet segmentation is exactly the same as the postings one. I only see this highlighter as a generalization of the postings highlighter. After some benchmarking it seems to perform quite well on plain analysis (twice faster than the original plain highlighter on wikipedia text) but the drawbacks are the same than for the original Postings Highlighter.

Ok then.....

javanna commented:
@jimczi were your benchmarks based on lucene code or elasticsearch code from this PR?

jimczi commented:
@javanna elasticsearch code from this PR. Well not exactly a benchmark though, only me playing with some queries ;)

clintongormley commented:
Does the unified highlighter support fragment length?  I tried it out on https://github.com/elastic/elasticsearch/issues/9442 and it gives better results than the existing highlighter, but does not appear to limit the length.

clintongormley commented:
Does this issue apply to the unified highlighter too? https://github.com/elastic/elasticsearch/issues/11223

clintongormley commented:
The unified highlighter doesn't seem to work with boundary characters but doesn't complain about them not being supported either. See https://github.com/elastic/elasticsearch/issues/11777

javanna commented:
If the code is the same as the postings highlighter, fragments look good with natural language but it does not support fragment size. One sentence per fragment is usually the output.

clintongormley commented:
Looks like the unified highlighter doesn't support `require_field_match: false`:

```
PUT t/t/1
{
  "foo": "one",
  "bar": "one two"
}

GET t/_search
{
  "query": {
    "match": {
      "_all": "one"
    }
  },
  "highlight": {
    "require_field_match": false,
    "type": "unified",
    "fields": {
      "*": {}
    }
  }
}
```

javanna commented:
> Looks like the unified highlighter doesn't support require_field_match: false

I think this is the same limitation that the postings highlighter has (since we moved to the standard lucene postings highlighter in 2.0).

clintongormley commented:
One thing the plain highlighter does better than the unified highlighter is to collapse contiguous highlights, eg if you have ngrams length 1, the UH will return `<b>a</b><b>b</b><b>c</b>` while the plain highlighter will return `<b>abc</b>`.  Perhaps this could be done in a simple post-processing step?

clintongormley commented:
@dsmiley compared the UH to wikimedia's experimental highlighter in https://github.com/wikimedia/search-highlighter/issues/19#issue-179526836 which answers some of my questions above.  My testing of the UH so far shows that it is awesome, and makes me want to rip the other highlighters out.  But there are a few things that it will need to support before we can do so:

* `require_field_match`
* `matched_fields` (merged fields)
* `fragment_length` (position gap based fragementation)
* `boundary_characters` (maybe)

Also if, like our current implementation of the postings highlighter, this implementation of UH only highlights one doc at once (https://github.com/elastic/elasticsearch/issues/11223), we should look at just passing it TopDocs for better performance.

jimczi commented:
I updated the description with the current status of this PR. Some features are still missing or currently implemented in Lucene. My focus right now is to fix the highlighting on _all which is broken in this PR.

> Looks like the unified highlighter doesn't support require_field_match: false:

`require_field_match` should work, what doesn't work is the highlighting on _all (see my comment above).

jimczi commented:
> Does this issue apply to the unified highlighter too? #11223

Yes, it can highlight multiple docs and fields in one pass. Though a change in the ES highlighter API is needed first:
https://github.com/elastic/elasticsearch/issues/11223

dsmiley commented:
Thanks for the mention @clintongormley to let me know of this issue/thread.  I really think the UH can handle most use-cases for highlighting, and those lingering identified issues above can be addressed in the UH to make it better.

There are some issues filed in Lucene already:
* http://issues.apache.org/jira/browse/LUCENE-7565  (fragsize char based fragmentation)
* None yet for the mergeFields FVH feature equivalent, but I figure it'll be at least filed before the year end.
* https://issues.apache.org/jira/browse/LUCENE-7575 requireFieldMatch=false.  IMO this is the biggest feature users need before converting to the UH if the UH doesn't already meet their needs.

The existence of any filed issue shouldn't imply I'm about to work on it.  It would be cool if some of the folks here might help get some of these done.  It's looking doubtful that I'll have time by year's end to do these myself.

The issue mentioned earlier about adjacent words getting separate tags -- this is something that could be addressed in a custom PassageFormatter (or enhance the DefaultPassageFormatter).  This might be a small hack that makes certain language assumptions as to what characters can be in-between (not alpha-numeric).  A more correct fix would be in the UH to expose some sort of clause-ID (an index distinguishing one OffsetsEnum from the next) to the Passage, and in addition doing phrases differently to produce one OffsetsEnum per clause instead of wrapping each word separately.  Some of that may happen for other reasons like improving passage relevancy of phrases/spans.

jimczi commented:
Thanks @dsmiley 
I though that requireFieldMatch was easy to implement outside of Lucene thanks to the extensibility of the UnifiedHighlighter but my implementation does not cover all the cases. I'll revert the partial implementation that I wrote for this PR and since you open http://issues.apache.org/jira/browse/LUCENE-7565 I pushed a patch on the issue to add this feature directly to Lucene. 
I also have some idea regarding http://issues.apache.org/jira/browse/LUCENE-7565 so I'll add my solution there if the implementation is smooth ;).

> and in addition doing phrases differently to produce one OffsetsEnum per clause instead of wrapping each word separately. Some of that may happen for other reasons like improving passage relevancy of phrases/spans.

I would love to see this. I think the remaining complexity of the beautiful and simple ( ;) ) UnifiedHighlither is the fact that it tries to highlight single terms instead of queries even when span/phrase queries are involved.

BTW thank you for this great highlighter, I look forward using it in ES ;).

dsmiley commented on core/src/main/java/org/apache/lucene/search/uhighlight/CustomUnifiedHighlighter.java:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    @Override
+    protected PhraseHelper getPhraseHelper(String field, Query query, EnumSet<HighlightFlag> highlightFlags) {
```
The PhraseHelper is likely to become package-private in a future LUCENE JIRA issue.  Why do you need access to it?

jimczi replied:
I though it was needed to implement `require_field_match` but it doesn't work as I expected.
I'll revert this part and rely on the implementation in http://issues.apache.org/jira/browse/LUCENE-7575.

jimczi commented on core/src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java:
```diff
@@ -87,21 +86,18 @@ public Query rewrite(IndexReader reader) throws IOException {
-            return new MatchNoDocsQuery();
-        }
+        // if the terms does not exist we could return a MatchNoDocsQuery but this would break the unified highlighter
+        // which rewrites query with an empty reader.
```
I did not find a better way to make this query works with the unified highlighter. It's a simple term query so I don't think that this will have any impact but it's not very clean ;)

nik9000 replied:
Is this just going to make queries again `_all` a little less efficient what `_all` is disabled? If so I think that is ok.

jimczi commented on core/src/main/java/org/elasticsearch/common/lucene/search/MultiPhrasePrefixQuery.java:
```diff
@@ -150,7 +166,12 @@ public Query rewrite(IndexReader reader) throws IOException {
+            return new BooleanQuery.Builder()
+                .add(query.build(), BooleanClause.Occur.MUST)
+                .add(Queries.newMatchNoDocsQuery("No terms supplied for " + MultiPhrasePrefixQuery.class.getName()),
+                    BooleanClause.Occur.MUST).build();
```
Same here, we return a query that returns no result but extractTerms will return the terms that are already expanded.

jimczi commented:
I rebased this PR, cleanup all the tests and added some documentation. I also removed my custom implementation of the `require_field_match` option. I think it's better to merge this PR first and then we can integrate the missing options since
`require_field_match` and `merged_fields` are handled by http://issues.apache.org/jira/browse/LUCENE-7575
I changed a bit some of the queries that were not correctly handled by this new highlighter. I left some comments on the part that I am not sure of.
@jpountz @nik9000 @javanna could you please take a look ?

dsmiley commented:
Unless I misunderstand what ES's `merge_fields` is, I don't believe LUCENE-7575 handles it.  AFAIK `merge_fields` would consume the positions from multiple fields and highlight one stored field.  A use-case is analyzing the same text differently, perhaps with stemming in one field but not another, and then only storing the text once.

jimczi commented:
> Unless I misunderstand what ES's merge_fields is, I don't believe LUCENE-7575 handles it.

That's correct. What I meant is that LUCENE-7575 is more than `require_field_match` and that it allows to select which field to extract from the query. I realized now that I mixed up `merged fields` and `matched_fields` that does what you describe. I'll update the description, thanks.

jimczi commented:
@elasticmachine retest this please

jimczi commented:
I pushed another iteration now that we upgraded to the latest Lucene release. The unified highlighter now supports `require_field_match`. I think it is enough to merge this in 5.x. 
@nik9000  can you take a look ?

jimczi commented on core/src/main/java/org/apache/lucene/search/uhighlight/CustomPassageFormatter.java:
```diff
@@ -0,0 +1,82 @@
+import org.elasticsearch.search.fetch.subphase.highlight.HighlightUtils;
+
+/**
+Custom passage formatter that allows us to:
```
We could move it to another package, the only reason to put it here is because we already have the custom postings highlighter in the `org.apache.lucene` package.

jimczi commented:
Thanks @nik9000 
I pushed some change to address your comments. 
Regarding the "rewrites with empty reader" thing, this is required because the unified highlighter uses the rewritten query to build the automaton for the re-analysis strategy.

jimczi commented:
Thanks @nik9000 !
Now merging to 5.x

fariaintrafind commented:
The current integration doesn't seem to support setting the `maxLength` field in UnifiedHighlighter. Is this correct, or did I miss something?

jimczi commented:
That's correct @fariaintrafind , the UnifiedHighlighter uses a sentence break iterator that does not take the `maxFragmentLength` in account. Though we plan to support this setting soon, it's on my todos ;)

lami02 commented:
I have tested the new highlighter and it works great so far. The only setting that does not seem to work (except from the settings that are not supported yet) is `no_match_size`. The `number_of_fragments` setting doesn't seem to do anything either, but I suspect thats because it is not possible to set a `fragment_size` yet.

bleskes commented:
@lami02 thanks for trying it out! do you mind opening a new issue with the errors you found? a concise reproduction with API calls will be great.
